### PR TITLE
[GPII-3363]: Filter sensitive data better, various fixes

### PR DESCRIPTION
**Pull request by natarajaya**
This should solve the issue with leaking TF state key during Terragrunt configuration phase.

### CODE REVIEW

**Comment by amatas** [comment]
LGTM

File: shared/rakefiles/sh_filter.rb [hunk 1]
```diff
@@ -2,15 +2,17 @@
 # In case target command terminates with non-zero exit status, it will exit() with the same
 # status, preventing further rake execution, unless exit_on_nonzero_status is set to false.
 def sh_filter(cmd, exit_on_nonzero_status = true)
-  process = IO.popen(cmd).each do |out|
+  # We need to merge stderr and stdout streams of child process to capture its output entirely:
+  # https://ruby-doc.org/core-2.5.1/Kernel.html#method-i-spawn
+  process = IO.popen(cmd, :err => [:child, :out]).each do |out|
```
[the remark this answers is not in the document]
Comment (natarajaya) [reply shared/rakefiles/sh_filter.rb]: @stepanstipl Why do you think we should treat both streams separately if the only purpose of this function is to make substitutions?

[the remark this answers is not in the document]
Comment (natarajaya) [reply shared/rakefiles/sh_filter.rb]: @stepanstipl I understand your point, however, it is significantly more difficult to treat both streams separately, mostly due to IO synchronization issues (even using non-blocking reads it is impossible to handle both streams with respect to each other within same execution thread). How about I keep current approach, but make it configurable?

**Comment by natarajaya** [comment]
@stepanstipl Please check out updated version!

@amatas I took the liberty to bring back `rake sh` task in replacement of `rake rake_shell`, while applying fixes to preserve stderr for interactive shell. It appears in some places of our documentation, it also got shorter name, which is much more convenient to use in development workflow. It can do plain shell with `rake sh[bash,plain]`.

**Comment by mrtyler** [comment]
@natarajaya 
> It can do plain shell with rake sh[bash,plain].

How will a new user figure this out? Is this pattern documented somewhere?

**Comment by amatas** [comment]
So to get a shell inside the container I need to do `rake sh[bash,plain]` instead of `rake plain_shell`. IMHO is more uncomfortable and also it is not easy to figure out.

**Comment by natarajaya** [comment]
@amatas @mrtyler Ok, I got your point guys. Added `plain_sh` task and updated descriptions with some relevant usage examples.

File: shared/rakefiles/entrypoint.rake [hunk 2]
```diff
@@ -108,22 +108,30 @@ task :unlock => [:set_vars] do
...
+end
+
+desc "[ADVANCED] Run arbitrary command in exekube container via plain shell -- rake plain_sh['kubectl --namespace gpii get pods']"
+task :plain_sh, [:cmd] => [:set_vars] do |taskname, args|
+  if args[:cmd]
+    cmd = args[:cmd]
   else
-    sh "#{@exekube_cmd} rake xk['#{cmd}',skip_secret_mgmt,preserve_stderr]"
+    puts "Argument :cmd -- the command to run inside the exekube container -- not present, defaulting to bash"
+    cmd = "vash"
```
Comment (mrtyler) [inline shared/rakefiles/entrypoint.rake]: Unless [this guy](https://en.wikipedia.org/wiki/Vash_the_Stampede) is a shell now, this might be a typo. ;)

**Comment by amatas** [comment]
LGTM

**Pull request merged**